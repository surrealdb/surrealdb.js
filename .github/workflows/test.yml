name: test

on:
  push:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Setup node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18
      - name: Setup deno
        uses: denoland/setup-deno@v1
        with:
          deno-version: v1.x
      - name: Setup bun
        uses: antongolub/action-setup-bun@v1
        with:
          bun-version: 0.1.13

      - uses: actions/checkout@v3

      - name: Build NPM package
        run: deno task build

      - name: Start SurrealDB
        run: sh ./test/e2e/scripts/docker-restart.sh

      - name: Run Deno
        run: deno test -A --trace-ops ./test/e2e/deno.js
      - name: Run Node
        run: node ./test/e2e/node.js
      - name: Run Bun
        run: bun run ./test/e2e/bun.js

      - name: Stop SurrealDB
        run:  sh ./test/e2e/scripts/docker-stop.sh
