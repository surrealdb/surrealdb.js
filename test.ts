import { RecordId } from "./src/library/data/recordid.ts";
import { Duration } from "./src/library/data/duration.ts";
import { UUID } from "./src/library/data/uuid.ts";
import { Decimal } from "./src/library/data/decimal.ts";
import { Surreal } from "./src/index.ts";
import { encode } from "./src/library/data/cbor.ts";
import { GeometryCollection, GeometryLine, GeometryPoint } from "./src/library/data/geometry.ts";
// import { recordId } from "./src/library/orm/types.ts";

const surreal = new Surreal();
await surreal.connect('ws://127.0.0.1:8000/rpc');
await surreal.use({
	namespace: 'test',
	database: 'test',
});

// console.log(
// 	await surreal.query<[unknown[]]>(
// 		"[$d, $r, $c, $u, $dur, $none, NONE, $a, <string> $a, $nan, $geo]",
// 		{
// 			d: new Date(),
// 			r: new RecordId('person', 'tobie'),
// 			c: new RecordId('recording', [
// 				new Date(),
// 				'London',
// 				new Uuid(),
// 				{
// 					temparature: new Decimal("10.00003")
// 				}
// 			]),
// 			u: new Uuid(),
// 			dur: new Duration('1d20m'),
// 			none: undefined,
// 			a: 9223372036854775807n,
// 			nan: NaN,
// 			geo: new GeometryCollection([
// 				new GeometryPoint([1, 2]),
// 				new GeometryLine([ new GeometryPoint([1, 2]), new GeometryPoint([3, 4]) ])
// 			])
// 		}
// 	)
// )

const uuid = new Duration("2w6d100ms");
console.log(
	456,
	await surreal.query<[UUID, string]>(
		"$uuid; <string> $uuid",
		{
			uuid
		}
	).then(([u, a]) => [u.toString(), a])
)

// function bufferToHex (buffer: ArrayBuffer) {
//     return [...new Uint8Array (buffer)]
//         .map (b => b.toString (16).padStart (2, "0"))
//         .join ("");
// }

// console.log(bufferToHex(encode(new RecordId('recording', [
// 	new Date(),
// 	'London',
// 	new Uuid(),
// 	{
// 		temparature: new Decimal("10.00003")
// 	}
// ]))))

// const rid = recordId('test');
// console.log(rid.parse(new RecordId('tesst', 123) ))

// console.log()
