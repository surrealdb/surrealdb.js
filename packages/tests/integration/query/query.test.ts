import { beforeAll, describe, expect, test } from "bun:test";
import { RecordId, surql } from "surrealdb";
import { setupServer } from "../__helpers__";

const { createSurreal } = await setupServer();

beforeAll(async () => {
    const surreal = await createSurreal();

    await surreal.query(/* surql */ `
		CREATE |foo:1..100| CONTENT { hello: "world" };
	`);
});

describe("query()", async () => {
    const surreal = await createSurreal();

    test("direct query", async () => {
        await surreal.query(`UPSERT hello:world CONTENT { hello: "world" }`);
    });

    test("collect query results", async () => {
        const [result] = await surreal.query(`RETURN hello:world.hello`).collect<["world"]>();

        expect(result).toEqual("world");
    });

    test("stream query results", async () => {
        const stream = surreal.query(`SELECT * FROM foo`).stream();

        let valueCount = 0;
        let doneCount = 0;
        let errorCount = 0;

        for await (const frame of stream) {
            if (frame.isValue<{ hello: string }>()) {
                expect(frame.value.hello).toEqual("world");
                valueCount++;
            } else if (frame.isDone()) {
                doneCount++;
            } else if (frame.isError()) {
                errorCount++;
            }
        }

        expect(valueCount).toEqual(100);
        expect(doneCount).toEqual(1);
        expect(errorCount).toEqual(0);
    });

    test("bound query", async () => {
        const record = new RecordId("hello", "world");
        const query = surql`
			RETURN ${record}
		`;

        const [result] = await surreal.query(query).collect<[RecordId]>();

        expect(result.equals(record)).toBeTrue();
    });
});
