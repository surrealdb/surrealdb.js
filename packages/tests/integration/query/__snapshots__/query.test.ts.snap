// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`query() inner query 1`] = `"RETURN $bind__1"`;

exports[`query() inner query 2`] = `
{
  "bind__1": RecordId {},
}
`;
